include/alpaka/mem/concepts/IDataSource.hpp

```diff
@@ -27,22 +27,27 @@ namespace alpaka::concepts
                   /// access may be invalid after moving the DataSource.
                   static_cast<bool>(t);
 
-                  // only if no reference is not defined, check if the access operator returns a value
-                  requires requires { typename T::reference; } || requires {
-                      { mut_t[vec] } -> std::same_as<typename T::value_type>;
-                  };
-                  requires requires { typename T::const_reference; } || requires {
-                      { const_t[vec] } -> std::same_as<typename T::value_type>;
-                  };
+                  // check multi-dimensional mutable access operator
+                  requires
+                      (requires { typename T::reference; } && requires {{ mut_t[vec] } -> std::same_as<typename T::reference>; }) ||
+                      (requires { typename T::value_type; } && requires {{ mut_t[vec] } -> std::same_as<typename T::value_type>; });
```

Comment:
```suggestion
                  requires
                      (requires { typename T::reference; } && requires {{ mut_t[vec] } -> std::same_as<typename T::reference>; }) ||
                      requires {{ mut_t[vec] } -> std::same_as<typename T::value_type>; });
```

We can write this, because I check already if IDataSource has the member type `T::value_type`. Also it is maybe a little bit missing leading, because a `IMdSpan` has also a `T::value_type` member type. Therefore the `value_type` is no criteria for the return type of the access operator.

This can be applied on all following requirement clauses.

Reply:
If I understand you correctly you pointed to the issue that a data source can have defines `value_type` and `reference` but the access operator is returning `value_type` what will be ok from the current concept description but thats not we intended.

I pushed an update which is now saying if there is `reference` defined the access operator must provide a reference and not the `value_type`.
May be code would look better if we use a trait `hasReferenceDefined<>` and `hasValueTypeDefined<>` instead of requires, but this is butifying and therefore I have not added it.